controllers/gce/utils/utils_test.go

```diff
@@ -0,0 +1,482 @@
+		}
+	}
+}
+
+func TestNewNamer(t *testing.T) {
+	cn, fn := "fooCN", "fooFN"
+
+	namer := NewNamer(cn, fn)
+	rcn := namer.GetClusterName()
+	if rcn != cn {
+		t.Errorf("returned %v but expected %v", rcn, cn)
+	}
+
+	rfn := namer.GetFirewallName()
+	if rfn != fn {
+		t.Errorf("returned %v but expected %v", rfn, fn)
+	}
+}
+
+func TestTruncate(t *testing.T) {
+	nameLenLimitValue := ""
+	for i := 0; i < nameLenLimit; i++ {
+		nameLenLimitValue = fmt.Sprintf("%s%s", nameLenLimitValue, "t")
+	}
+
+	fooTests := []struct {
+		k  string
+		ek string
+	}{
+		{"", ""},
+		{"fooKey", "fooKey"},
+		{nameLenLimitValue + "more", nameLenLimitValue + "0"},
+	}
+
+	for _, fooTest := range fooTests {
+		namer := NewNamer("fooCN", "fooFN")
+		rk := namer.Truncate(fooTest.k)
+		if rk != fooTest.ek {
+			t.Errorf("returned %v but expected %v", rk, fooTest.ek)
+		}
+	}
+}
+
+func TestDecorateName(t *testing.T) {
+	nameLenLimitValue := ""
+	for i := 0; i < nameLenLimit-6; i++ {
+		nameLenLimitValue = fmt.Sprintf("%s%s", nameLenLimitValue, "t")
+	}
+
+	fooTests := []struct {
+		cn  string
+		n   string
+		edn string
+	}{
+		{"", "", ""},
+		{"", "fooName", "fooName"},
+		{"fooCN", "fooName", "fooName--fooCN"},
+		{"fooCN", nameLenLimitValue, nameLenLimitValue + "--fooC0"},
+	}
+
+	for _, fooTest := range fooTests {
+		namer := NewNamer(fooTest.cn, "fooFN")
+		rdn := namer.decorateName(fooTest.n)
+		if rdn != fooTest.edn {
+			t.Errorf("returned %v but expected %v", rdn, fooTest.edn)
+		}
+	}
+}
+
+func TestParseName(t *testing.T) {
+	fooTests := []struct {
+		n   string
+		ecn string
+		ers string
+	}{
+		{"", "", ""},
+		{"foo", "", ""},
+		{"foo01--", "", ""},
+		{"foo01--foo02", "foo02", ""},
+		{"foo01-foo02--foo03", "foo03", "foo02"},
+	}
+
+	namer := NewNamer("fooCN", "fooFN")
+	for _, fooTest := range fooTests {
+		nc := namer.ParseName(fooTest.n)
+		if nc == nil {
+			t.Errorf("returned nil")
+			continue
+		}
+
+		if nc.ClusterName != fooTest.ecn {
+			t.Errorf("returned %v but expected %v", nc.ClusterName, fooTest.ecn)
+			continue
+		}
+
+		if nc.Resource != fooTest.ers {
+			t.Errorf("returned %v but expected %v", nc.Resource, fooTest.ers)
+			continue
+		}
+	}
+}
+
+func TestNameBelongsToCluster(t *testing.T) {
+	nameLenLimitValue := "k8s-foo01--"
+	for i := 0; i < nameLenLimit; i++ {
+		nameLenLimitValue = fmt.Sprintf("%s%s", nameLenLimitValue, "t")
+	}
+	nn := &Namer{}
+	trn := nn.Truncate(nameLenLimitValue)
+
+	fooTests := []struct {
+		n  string
+		cn string
+		ev bool
+	}{
+		{"", "", false},
+		{"foo01", "", false},
+		{"k8s-foo01", "", true},
+		{"k8s-foo01", "foo01", false},
+		{"k8s-foo01--foo02", "", false},
+		{"k8s-foo01--foo02", "foo", false},
+		{"k8s-foo01--foo02", "foo02", true},
+		{"k8s-foo01--foo02", "k8s-foo01--foo02", true},
+		{"k8s-foo01--foo02--foo03", "foo02", false},
+		{"k8s-foo01--foo02--foo03", "foo02--foo03", false},
+		{"k8s-foo01--foo02--foo03", "k8s-foo01--foo02--foo03", false},
+		{trn, trn, true},
```

Comment:
Why do you have the truncated name as expected cluster name? 

```go
func TestNameBelongsToCluster(t *testing.T) {
	cluster := "560b5b8154db29b3"
	nameLenLimitValue := "k8s-tps-e2e-tests-ingress-upgrade-xxmpf-static-ip--" + cluster
	nn := &Namer{}
	trn := nn.Truncate(nameLenLimitValue)

	fooTests := []struct {
		n  string
		cn string
		ev bool
	}{
		{trn, cluster, true},
	}

	namer := NewNamer("fooCN", "fooFN")
	for _, fooTest := range fooTests {
		namer.SetClusterName(fooTest.cn)
		rv := namer.NameBelongsToCluster(fooTest.n)
		if rv != fooTest.ev {
			t.Errorf("returned %v but expected %v", rv, fooTest.ev)
		}
	}
}
```
This test uses a cluster name and resource name from an e2e test.  However, this test fails because there's a bug in `NameBelongsToCluster`